[AutoDiff] Always thunk JVP/VJPs during SILGen.

dan-zheng opened the pull request:
Always thunk JVPs/VJPs during SILGen.

- If actual JVP/VJP SIL function types match the expected JVP/VJP types
  (computed from the original function), create a simple forwarding thunk.
- Otherwise, perform reabstraction, handling parameter convention mismatches.
  - To convert an owned parameter to be guaranteed, retain it in the thunk.
  - To convert a guaranteed parameter to be owned, release it in the thunk.
  - Perform reabstraction in differential/pullback thunks in addition to
    self reordering. 
- This simplifies TBDGen: JVP/VJP symbols are always emitted.

Resolves [TF-546](https://bugs.swift.org/projects/TF/issues/TF-546), [TF-667](https://bugs.swift.org/projects/TF/issues/TF-667), and [TF-682](https://bugs.swift.org/projects/TF/issues/TF-682).

dan-zheng committed:
[AutoDiff] Fix JVP/VJP SILGen thunking logic.

- Create JVP/VJP thunks during SILGen whenever actual JVP/VJP types do not
  match expected types.
- Handle JVP/VJP parameter convention mismatch during SILGen using
  reabstraction thunks.
  - To convert an owned parameter to be guaranteed, retain it in the thunk.
  - To convert a guaranteed parameter to be owned, release it in the thunk.
- Generate reabstraction thunks for differentials/pullbacks.
  - Progress towards TF-625: maximally indirect differentials/pullbacks.

Resolves TF-667 and TF-682.

dan-zheng commented on test/AutoDiff/leakchecking.swift:
```diff
@@ -208,6 +208,116 @@ LeakCheckingTests.testWithLeakChecking("ClassMethods") {
+  testLinearMapSILGenThunks()
+}
+
+LeakCheckingTests.testWithLeakChecking("ParameterConventionMismatchLeakChecking") {
```
Note: here are leak checking tests for [TF-667](https://bugs.swift.org/projects/TF/issues/TF-667).

dan-zheng commented on test/AutoDiff/leakchecking.swift:
```diff
@@ -208,6 +208,116 @@ LeakCheckingTests.testWithLeakChecking("ClassMethods") {
+}
+
+// Test leaks regarding `SILGenFunction::getOrCreateAutoDiffLinearMapThunk`.
+LeakCheckingTests.testWithLeakChecking("LinearMapSILGenThunks") {
```
Note: here are leak checking tests for [TF-682](https://bugs.swift.org/projects/TF/issues/TF-682).

dan-zheng committed:
Add TBDGen tests, expose TDBGen failures.

From test/AutoDiff/tbdgen.swift:

<unknown>:0: error: symbol 'AD__$sSf4testE6method1x1yS2f_SftF__jvp_src_0_wrt_0_1_2' (AD__$sSf4testE6method1x1yS2f_SftF__jvp_src_0_wrt_0_1_2) is in TBD file, but not in generated IR
<unknown>:0: error: symbol 'AD__$sSf4testEyS2fcig__vjp_src_0_wrt_0_1' (AD__$sSf4testEyS2fcig__vjp_src_0_wrt_0_1) is in TBD file, but not in generated IR

rxwei commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -3397,112 +3399,274 @@ SILGenFunction::getOrCreateAutoDiffLinearMapReorderingThunk(
+        SGF.B.createRetainValue(loc, arg.getValue(),
+                                SGF.B.getDefaultAtomicity());
+      else
+        SGF.B.createRetainValueAddr(loc, arg.getValue(),
```
`retain_value_addr` can only be used on buffers of loadable types. I can't find any tests in this PR that involves abstraction differences in address-only parameters. If you write such an example I'm sure it will trigger an assertion.

When you convert an indirect address-only owned parameter (`@in`) to `@in_guaranteed`, you should allocate a new local buffer and perform `copy_addr` to create a copy of the underlying value. You can follow how this is done in the differentiation transform:
https://github.com/apple/swift/blob/b2720a3460c6eec5351e40f5fa2ddb95dd2df9a7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1956

dan-zheng replied:
Parameter convention conversions for loadable-address and address-only types done in https://github.com/apple/swift/pull/26380/commits/6cbc71765c33fd4fde0d2b0c0235be5496c129e9, thanks!

I [added tests](https://github.com/apple/swift/pull/26380/commits/6cbc71765c33fd4fde0d2b0c0235be5496c129e9#diff-5240fec94d5989e4aacc603204b9ac0c) involving `__shared`/`__owned`/`__consuming` mismatches for struct methods/initializers. Is this sufficient coverage, or is more testing necessary (protocol methods with/without `__consuming`)?

rxwei reviewed (commented):
Reabstraction for address-only parameters are not tested at all as far as I can tell. Please add tests and fix the underlying issues.

rxwei commented:
To be concrete, what you should test is converting between functions with `__shared` and `__owned` on generic parameters, and protocol methods with and without `__consuming`.

dan-zheng committed:
Extend conversion logic for parameter convention mismatches.

Handle conversions for loadable-address and address-only types.
Add tests.

rxwei commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -3588,13 +3741,41 @@ SILGenModule::getOrCreateAutoDiffAssociatedFunctionReorderingThunk(
+    if (arg->getType().isObject())
+      thunkSGF.B.emitReleaseValueAndFold(loc, arg);
+    else
+      thunkSGF.B.emitDestroyAddrAndFold(loc, arg);
```
Here `release_value` and `destroy_addr` are the correct instructions to use, but the "and fold" part is never exercised in thunks like this. I'd suggest using builder functions `createReleaseValue` and `createDestroyAddr` instead.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26380/commits/9d4bf257c138125575b97b2e2585f96376d55d85, thanks!

dan-zheng committed:
Always generate JVP/VJP thunks during SILGen.

If a user specifies JVP/VJP functions in a `@differentiable` attribute,
always generate a forwarding thunk for them during SILGen.

Previously, thunks were generated only when the actual JVP/VJP type did not
match the expected JVP/VJP type.

This simplifies TBDGen: JVP/VJP symbols should always be emitted.
For every `@differentiable` attribute:
- If JVP/VJP is specified, thunks are generated during SILGen.
- If JVP/VJP is not specified, the differentiation transform generates them.

rxwei commented on test/AutoDiff/leakchecking.swift:
```diff
@@ -215,6 +215,156 @@ LeakCheckingTests.testWithLeakChecking("ClassMethods") {
+    }
+
+    @differentiable(vjp: vjpConsuming)
+    __consuming func consuming(_ x: Tracked<Float>) -> Tracked<Float> {
```
Also add tests where the original is non-`__consuming` but the JVP and VJP are.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26380/commits/bba4b634e7ad23a0ce165746cfd074a66ad7ceb1.

dan-zheng committed:
Test indirect `Self`.

Change test/AutoDiff/leakchecking.swift to test indirect `Self` type.
Fix test/AutoDiff/subset_parameters_thunk.swift.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix `@differentiable` attribute serialization, add TF-546 test.

Serialize JVP/VJP names and function separately.
Add test for TF-546, which is fixed by reabstraction thunking.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
Thanks @rxwei for the thorough review 🙂

rxwei commented:
Now that everything is thunked, you can simply get rid of the VJP/JVP name and the function pointer from SILDifferentiableAttr.
